neuromation/cli/job.py

```diff
@@ -877,6 +873,55 @@ def format_fail(job: str, reason: Exception) -> str:
     return job
 
 
+async def _build_volumes(
+    root: Root, input_volumes: Sequence[str], env_dict: Dict[str, str]
+) -> Set[Volume]:
+    VOLUME_SHORTCUTS = {"HOME", "ALL"}
+    detected_shortcuts = VOLUME_SHORTCUTS & set(input_volumes)
+    if len(detected_shortcuts) > 1:
+        msg = ", ".join(f"--volume={shortcut}" for shortcut in detected_shortcuts)
+        raise click.UsageError(f"Cannot use multiple volume shortcuts together: {msg}")
+
+    volumes: Set[Volume] = set()
+    for vol in input_volumes:
+        if vol in VOLUME_SHORTCUTS:
+            if vol == "HOME":
+                volumes.add(
+                    root.client.parse.volume("storage://~:/var/storage/home:rw")
+                )
+                volumes.add(
+                    root.client.parse.volume(
+                        "storage://neuromation/public:/var/storage/neuromation:ro"
+                    )
+                )
+            if vol == "ALL":
+                available = await root.client.users.get_acl(
+                    root.username, scheme="storage"
+                )
+                root_mountpoint = "/var/storage"
+                env_dict[ENV_VAR_STORAGE_ROOT_MOUNTPOINT] = root_mountpoint
+                if not root.quiet:
+                    click.echo(
+                        "Storage root mountpoint will be available as the environment "
+                        f"variable: {ENV_VAR_STORAGE_ROOT_MOUNTPOINT}={root_mountpoint}"
+                    )
+                for perm in available:
+                    root_dir = (
+                        perm.uri.host if perm.uri.host != root.username else "home"
```

Comment:
not sure the `home` link is what we want really. I'd rather have a separate link (an additional mount) instead of renaming the owners home dir. also, we could easily introduce an environment var to address this (e.g. `$HOME`).

Reply:
In this PR, we made this decision to mount user home directory to `/var/storage/home` not to `/var/storage/username` in order to be compliant with `--volume=HOME` in order not to confuse users. I don't think we need to have both mounts.

Reply:
In addition, in a scheme `storage://username/path/to/dir`, we do not use `username` as a location of user's home directory, technically, in NFS it can be anywhere (for example, not in `/username` but `/home/username`), so I think it's good practice to mount user dir to `/var/storage/home`.

Reply:
I think this is wrong as this is not corresponding to what `neuro ls storage:/` returns.

Reply:
this is question of taste, let's ask our users (cc @mariyadavydova).

Currently:
- `--volume=HOME`: `storage://alice/` -> `/var/storage/home/`

We need to implement `--volume=ALL`. Suppose, `alice` is using this command. There are 2 options:

1. replace user's name with `home` in order to be consistent with `--volume=HOME` (and not cause surprises for our users):
- `storage://alice/dir1` -> `/var/alice/dir1`
- `storage://bob/dir2` -> `/var/bob/dir2`

2. as @dalazx proposes, do not change anything and thus simplify algorithm for `--volume=ALL`:
- `storage://alice/dir1` -> `/var/alice/dir1`
- `storage://bob/dir2` -> `/var/bob/dir2`

Reply:
What if there is a user "home"?

Reply:
there won't be any surprises as it is `ALL` and not `HOME`. we are not changing `HOME` behavior.

I think we should indicate a user's home directory via an env var (e.g `$HOME`).

another option is to mount the tree as is, but add an additional mount (a sort of a link) for `home`.

Reply:
> What if there is a user "home"?

if we chose the single mount, this is fine (in both cases, either we rename username to `home` or not, it's not a problem). The problem will appear if we choose to mount both.

Reply:
> we are not changing HOME behavior.

true.

> there won't be any surprises as it is ALL and not HOME

let's wait for our user's opinion.

Reply:
@ayushkovskiy I am your user as well. I want to try to run a file server and do not want to mess around some weird directories such as `home`, but rather want to get the same hierarchy as our storage API exposes.

Reply:
So do we agree on the following:
- ` --volume=HOME` mounts user's home directory to `/var/storage/home`, env var: `NEUROMATION_STORAGE_ROOT=/var/storage`
- `--volume=ALL` mounts everything available to `/var/storage` (including user's home dir mounted to `/var/storage/{username}`), env vars: `NEUROMATION_STORAGE_ROOT=/var/storage`, `NEUROMATION_STORAGE_HOME=/var/storage/{username}`

Reply:
The home directory is mounted as `/var/storage/home` and available under `${NEUROMATION_STORAGE_ROOT}/home` if `--volume=HOME`.
At the same time it is mounted as `/var/storage/{username}` and available under `${NEUROMATION_STORAGE_HOME` if `--volume=ALL`.

Does anybody feel a little inconsistency?

Reply:
@asvetlov agree it might seem inconsistent.
`HOME` also mounts `"storage://neuromation/public:/var/storage/neuromation:ro"` which is even more inconsistent.
I think that we just need to distinguish the two variants properly.
`HOME` - happens to be a set of hardcoded mounts.
`ALL` - a tree similar to what storage API exposes.

re your concern, should we not use `$NEUROMATION_STORAGE_ROOT` for `HOME`?

I think the goal here is not to introduce any regressions at least.

Reply:
Hard question.
Wild idea: we can keep existing mounts for backward compatibility *and* provide them in a consistent way under different path, e.g. `/var/neuro`. 
`/var/neuro` is a storage root, `/var/neuro/asvetlov` becomes my home etc. 
Thus `--volume=HOME` becomes a subset of `--volume=ALL` (plus legacy `/var/storage` mount which will be removed eventually, say in 6 months).

The downside is the absence of an ability to show deprecation warning in `neuro run` command, only in docs.

Reply:
brilliant!

Reply:
> Thus --volume=HOME becomes a subset of --volume=ALL (plus legacy /var/storage mount which will be removed eventually, say in 6 months).

@asvetlov do you mean with `--volume=ALL` we mount everything unchanged to `/var/neuro` (without mounting anything to `/var/storage`), and with `--volume=HOME` we mount all as-is?

> The downside is the absence of an ability to show deprecation warning in neuro run command, only in docs.

Why not? I don't see :(

Reply:
With `--volume=HOME` we mount both `/var/storage/home` and `/var/neuro/asvetlov`, the same for public: `/var/storage/neuromation` and `/var/neuro/neuromation/public`.

Reply:
I think we can keep `HOME` as is, but address the inconsistency in `ALL` as @asvetlov suggested if we ever decide to drop `HOME`.

Reply:
Yes, this is an option. If we decide to not touch HOME but remove it entirely somewhen we can add a deprecation now.
Is there a reason for mounting home and neuromation public only but skipping all other shared resources? Is `--volume=HOME` useful at all if we have `--volume=ALL` implemented?

Reply:
not sure. I guess we'll find out once we onboard more people. :)

Reply:
Agree

Reply:
OK thanks everyone for sharing your opinions, here's the final version:
- for `--volume=HOME`, we don't change anything (just print a deprecation warning)
- for `--volume=ALL`, we mount everything available (by permissions) to `/var/neuro` and expose 2 variables: `NEUROMATION_ROOT=/var/neuro`, `NEUROMATION_HOME=/var/neuro/{username}`

we also forbid both `--volume=ALL` and any other `--volume` option.

Reply:
I like the final version, thank you!

Considering HOME vs ALL: I think that the usage of these options is merely a question of examples we provide. Somehow I completely missed the HOME option being available. I assume that many existing users may be unaware of it because they are used to mounting particular URIs and PATHs. So we can actively promote ALL and never speak about HOME, and deprecate it soon.

I have a question though: from the technical point of view, if one have access to many shared resources, is it harder to mount ALL than to mount 2-3 paths? I mean can we safely recommend users to use ALL, or is there some trade off?

Reply:
In general I'd rather promote specific mounts rather than `ALL` to embrace isolation. though a user should be informed that there's an option to mount the whole available tree if needed.

I remember an interesting point @truskovskiyk brought when he was working on Hokutake. Kyryl did not even want `HOME` as he was afraid to corrupt the dataset and weights for the project by chance.

Reply:
Indeed, that's a very good point. Corrupting the dataset or the weights may be extremely expensive accident.

All right, I will stick to recommending mounting specific URIs as a best practice in our tutorials and receipts, while providing a information about ALL shortcut for something like quick experiments or data exploration.

Reply:
> as he was afraid to corrupt the dataset and weights for the project by chance

How about we introduce (in future) an option or a syntax extension to not mount or mount in read-only mode some specific important directories? Though now it's hard for me to propose any such syntax

Reply:
@ayushkovskiy This is a very good idea! We can introduce something like this:

```
> neuro storage protect storage:project/data
> neuro run ... --volume storage:project/data:/var/storage/data:rw ...
Warning: You are trying to mount storage:project/data with read-write access, while this directory is marked as protected. Access us downgraded to read-only. To get read-write access for this directory please remove the protection mode with
neuro storage protect --remove storage:project/data
```

Reply:
looks too complicated in my opinion.

Reply:
I think it should be an option for "serious teams, concerned about data security". I mean, in most cases you can simply watch out your commands and use proper access modifiers. But if you are **really** concerned, you may additionally use such an option as a safety net. This may be especially relevant for teams.

Reply:
I agree we should print a warning for all `rw` volumes

Reply:
https://github.com/neuromation/platform-client-python/issues/1018